File: internal/driver/codex.go
package driver

import (
	"encoding/json"
	"fmt"
	"strings"
)

// Codex implements the Driver interface for codex CLI.
type Codex struct{}

func (c *Codex) Name() string   { return "codex" }
func (c *Codex) Binary() string { return "codex" }

func (c *Codex) Models() []string {
	return []string{
		"gpt-5.3-codex",
		"gpt-5.2-codex",
		"gpt-5.1-codex-max",
		"gpt-5.1-codex",
		"gpt-5-codex",
		"gpt-5-codex-mini",
	}
}

func (c *Codex) HeadlessArgs(prompt, model, dir string) []string {
	var args []string
	if dir != "" {
		args = append(args, "-C", dir)
	}
	args = append(args, "exec", prompt, "--json", "--full-auto")
	if model != "" {
		args = append(args, "--model", model)
	}
	return args
}

func (c *Codex) InteractiveArgs(_ string, _ string, context, model string) []string {
	args := []string{"-a", "never", context}
	if model != "" {
		args = append(args, "--model", model)
	}
	return args
}

func (c *Codex) ResumeArgs(sessionID, feedback, model, dir string) []string {
	var args []string
	if dir != "" {
		args = append(args, "-C", dir)
	}
	args = append(args, "exec", "resume", sessionID, feedback, "--json", "--full-auto")
	if model != "" {
		args = append(args, "--model", model)
	}
	return args
}

func (c *Codex) ParseEvent(line []byte) (Event, error) {
	var raw map[string]any
	if err := json.Unmarshal(line, &raw); err != nil {
		return Event{Type: EventUnknown, Raw: string(line)}, nil
	}

	typ, _ := raw["type"].(string)

	if typ == "item.completed" {
		if item, ok := raw["item"].(map[string]any); ok {
			if itemType, _ := item["type"].(string); itemType == "agent_message" {
				text, _ := item["text"].(string)
				return Event{Type: EventText, Text: text, Raw: string(line)}, nil
			}
		}
	}

	if typ == "turn.completed" {
		if usage, ok := raw["usage"].(map[string]any); ok {
			cost := &Cost{}
			if in, ok := usage["input_tokens"].(float64); ok {
				cost.InputTokens = int64(in)
			}
			if out, ok := usage["output_tokens"].(float64); ok {
				cost.OutputTokens = int64(out)
			}
			return Event{Type: EventCost, Cost: cost, Raw: string(line)}, nil
		}
	}

	if typ == "thread.started" {
		if threadID, ok := raw["thread_id"].(string); ok {
			return Event{Type: EventSession, SessionID: threadID, Raw: string(line)}, nil
		}
	}

	return Event{Type: EventStatus, Raw: string(line)}, nil
}

func (c *Codex) ParseSessionID(events []Event) string {
	for _, e := range events {
		if e.SessionID != "" {
			return e.SessionID
		}
	}
	return ""
}

func (c *Codex) FormatEvent(line []byte) string {
	var raw map[string]any
	if err := json.Unmarshal(line, &raw); err != nil {
		return ""
	}

	typ, _ := raw["type"].(string)
	item, _ := raw["item"].(map[string]any)

	switch typ {
	case "item.started":
		if item == nil {
			return ""
		}
		itemType, _ := item["type"].(string)
		switch itemType {
		case "command_execution":
			cmd, _ := item["command"].(string)
			if cmd == "" {
				return ""
			}
			return fmt.Sprintf("[command: %s]\n", cmd)
		case "file_change":
			file, _ := item["file"].(string)
			if file == "" {
				return "[file]\n"
			}
			return fmt.Sprintf("[file: %s]\n", file)
		case "mcp_tool_call":
			name, _ := item["name"].(string)
			if name == "" {
				return "[mcp]\n"
			}
			return fmt.Sprintf("[mcp: %s]\n", name)
		case "reasoning":
			return "[thinking...]\n"
		}
		return ""
	case "item.completed":
		if item == nil {
			return ""
		}
		itemType, _ := item["type"].(string)
		switch itemType {
		case "agent_message":
			text, _ := item["text"].(string)
			if text == "" {
				return ""
			}
			if strings.HasSuffix(text, "\n") {
				return text
			}
			return text + "\n"
		case "reasoning":
			text, _ := item["text"].(string)
			if text == "" {
				return ""
			}
			return fmt.Sprintf("[thinking] %s\n", text)
		case "command_execution":
			if code, ok := item["exit_code"].(float64); ok && code != 0 {
				return fmt.Sprintf("[exit: %d]\n", int(code))
			}
		case "file_change":
			file, _ := item["file"].(string)
			if file != "" {
				return fmt.Sprintf("[file: %s]\n", file)
			}
		}
		return ""
	case "turn.completed":
		if usage, ok := raw["usage"].(map[string]any); ok {
			var in, out int64
			if v, ok := usage["input_tokens"].(float64); ok {
				in = int64(v)
			}
			if v, ok := usage["output_tokens"].(float64); ok {
				out = int64(v)
			}
			if in > 0 || out > 0 {
				return fmt.Sprintf("[tokens: %s in / %s out]\n", formatTokens(in), formatTokens(out))
			}
		}
		return ""
	case "turn.started":
		return "[turn started]\n"
	case "turn.failed":
		if errMsg, ok := raw["error"].(string); ok && strings.TrimSpace(errMsg) != "" {
			return fmt.Sprintf("[error: %s]\n", errMsg)
		}
		if errObj, ok := raw["error"].(map[string]any); ok {
			if msg, _ := errObj["message"].(string); strings.TrimSpace(msg) != "" {
				return fmt.Sprintf("[error: %s]\n", msg)
			}
		}
		return "[error]\n"
	case "thread.started":
		if tid, ok := raw["thread_id"].(string); ok && tid != "" {
			return fmt.Sprintf("[session: %s]\n", tid)
		}
		return ""
	case "item.updated":
		if item == nil {
			return ""
		}
		itemType, _ := item["type"].(string)
		status, _ := item["status"].(string)
		if status != "" {
			return fmt.Sprintf("[%s: %s]\n", itemType, status)
		}
		return ""
	default:
		// Never silently drop — show the event type for anything unrecognized
		return fmt.Sprintf("[%s]\n", typ)
	}
}

File: server/src/plugin/codex.ts
import type {
  HeadlessOpts,
  InteractiveOpts,
  ToolPlugin,
  ToolPluginEvent,
} from './types';

async function codexHeadlessArgs(
  prompt: string,
  model: string,
  dir: string,
  opts?: HeadlessOpts,
): Promise<string[]> {
  const args = ['exec', prompt, '--json', '--full-auto'];
  const resolvedDir = dir.trim();
  const resolvedModel = model.trim();
  if (resolvedDir) {
    args.unshift(resolvedDir);
    args.unshift('-C');
  }
  if (resolvedModel) {
    args.push('--model', resolvedModel);
  }
  if (opts?.schemaPath) {
    args.push('--output-schema', opts.schemaPath);
  } else if (opts?.jsonSchema) {
    const path = await writeSchemaFile(dir, opts.jsonSchema);
    args.push('--output-schema', path);
  }
  return args;
}

async function codexResumeArgs(
  sessionID: string,
  feedback: string,
  model: string,
  dir: string,
  _opts?: HeadlessOpts,
): Promise<string[]> {
  const args = ['exec', 'resume', sessionID, feedback, '--json', '--full-auto'];
  const resolvedDir = dir.trim();
  const resolvedModel = model.trim();
  if (resolvedDir) {
    args.unshift(resolvedDir);
    args.unshift('-C');
  }
  if (resolvedModel) {
    args.push('--model', resolvedModel);
  }
  return args;
}

async function codexInteractiveArgs(opts: InteractiveOpts): Promise<string[]> {
  const args = ['--full-auto'];
  if (opts.model.trim()) {
    args.push('--model', opts.model.trim());
  }
  if (opts.systemPrompt.trim()) {
    args.push(
      '-c',
      `developer_instructions=${JSON.stringify(opts.systemPrompt.trim())}`,
    );
  }
  return args;
}

function codexParseEvent(line: Buffer): ToolPluginEvent | null {
  const raw = line.toString('utf8').replaceAll('\r', '').trim();
  if (!raw) return null;

  const parsed = parseJSON(raw) as Record<string, unknown> | null;
  if (!parsed || typeof parsed !== 'object') {
    return { type: 'text', text: raw, raw };
  }

  const item = parsed.item as Record<string, unknown> | undefined;
  const type = asString(parsed.type);
  if (type === 'thread.started') {
    const sessionID = asString(parsed.thread_id);
    if (!sessionID) return null;
    return { type: 'session', sessionID, raw };
  }

  if (type === 'item.started') {
    const itemType = asString(item?.type);
    if (itemType !== 'mcp_tool_call') return null;
    return {
      type: 'tool_use',
      toolName: asString(item?.name) || 'tool',
      toolInput: (item?.input as Record<string, unknown>) ?? {},
      toolUseID: asString(item?.id),
      raw,
    };
  }

  if (type === 'item.completed') {
    const itemType = asString(item?.type);
    if (itemType === 'agent_message') {
      const text = pickMessageText(item);
      if (!text) return null;
      const separated = text.endsWith('\n') ? text : `${text}\n`;
      return { type: 'text', text: separated, raw };
    }
    if (itemType === 'mcp_tool_call') {
      return {
        type: 'tool_result',
        toolName: asString(item?.name),
        toolResult:
          (item?.output as Record<string, unknown>) ??
          (item?.result as Record<string, unknown>) ??
          {},
        toolUseID: asString(item?.id),
        isError: Boolean(item?.is_error),
        raw,
      };
    }
    return null;
  }

  return null;
}

function codexParseSessionID(events: ToolPluginEvent[]): string | null {
  for (let i = events.length - 1; i >= 0; i -= 1) {
    const event = events[i];
    if (event?.type !== 'session') continue;
    const session = event.sessionID?.trim();
    if (session) return session;
  }
  return null;
}

export const codexPlugin: ToolPlugin = {
  name: () => 'codex',
  binary: () => 'codex',
  models: () => [
    'gpt-5.4',
    'gpt-5.3-codex',
    'gpt-5.2-codex',
    'gpt-5.1-codex-max',
    'gpt-5.1-codex',
    'gpt-5-codex',
    'gpt-5-codex-mini',
  ],
  headlessArgs: codexHeadlessArgs,
  resumeArgs: codexResumeArgs,
  interactiveArgs: codexInteractiveArgs,
  parseEvent: codexParseEvent,
  parseSessionID: codexParseSessionID,
};

function pickMessageText(item: any): string {
  if (typeof item?.text === 'string' && item.text.trim())
    return item.text.trim();
  const content = item?.content;
  if (typeof content === 'string') return content.trim();
  if (!Array.isArray(content)) return '';
  const textParts: string[] = [];
  for (const part of content) {
    if (!part || typeof part !== 'object') continue;
    if (typeof part.text === 'string' && part.text.trim()) {
      textParts.push(part.text.trim());
      continue;
    }
    if (typeof part.content === 'string' && part.content.trim()) {
      textParts.push(part.content.trim());
    }
  }
  return textParts.join('\n').trim();
}

function parseJSON(raw: string): any | null {
  try {
    return JSON.parse(raw);
  } catch {
    return null;
  }
}

function asString(value: unknown): string {
  return typeof value === 'string' ? value.trim() : '';
}

async function writeSchemaFile(
  dir: string,
  schema: Record<string, unknown>,
): Promise<string> {
  const orcaDir = `${dir.replace(/\/+$/, '')}/.orca`;
  await Bun.$`mkdir -p ${orcaDir}`;
  const schemaPath = `${orcaDir}/step-schema.json`;
  await Bun.write(schemaPath, JSON.stringify(schema, null, 2) + '\n');
  return schemaPath;
}
